br/pkg/lightning/backend/external/file.go

```diff
@@ -0,0 +1,187 @@
+// given dataWriter and range properties will be maintained in the given
+// rangePropertiesCollector.
+func NewKeyValueStore(
+	ctx context.Context,
+	dataWriter storage.ExternalFileWriter,
+	rangePropertiesCollector *rangePropertiesCollector,
+	writerID int,
+	seq int,
+) (*KeyValueStore, error) {
+	kvStore := &KeyValueStore{
+		dataWriter: dataWriter,
+		ctx:        ctx,
+		rc:         rangePropertiesCollector,
+		writerID:   writerID,
+		seq:        seq,
+		u64Buffer:  make([]byte, 8),
+	}
+	return kvStore, nil
+}
+
+// AddKeyValue saves a key-value pair to the KeyValueStore. If the accumulated
+// size or key count exceeds the given distance, a new range property will be
+// appended to the rangePropertiesCollector with current status.
+func (s *KeyValueStore) AddKeyValue(key, value []byte) error {
+	kvLen := len(key) + len(value) + 16
+
+	// data layout: keyLen + key + valueLen + value
+	_, err := s.dataWriter.Write(
+		s.ctx,
+		binary.BigEndian.AppendUint64(s.u64Buffer[:0], uint64(len(key))),
+	)
+	if err != nil {
+		return err
+	}
+	_, err = s.dataWriter.Write(s.ctx, key)
+	if err != nil {
+		return err
+	}
+	_, err = s.dataWriter.Write(
+		s.ctx,
+		binary.BigEndian.AppendUint64(s.u64Buffer[:0], uint64(len(value))),
+	)
+	if err != nil {
+		return err
+	}
+	_, err = s.dataWriter.Write(s.ctx, value)
+	if err != nil {
+		return err
+	}
+
+	if len(s.rc.currProp.key) == 0 {
+		s.rc.currProp.key = key
+		s.rc.currProp.writerID = s.writerID
+		s.rc.currProp.dataSeq = s.seq
+	}
+
+	s.offset += uint64(kvLen)
+	s.rc.currProp.size += uint64(len(key) + len(value))
+	s.rc.currProp.keys++
+
+	if s.rc.currProp.size >= s.rc.propSizeIdxDistance ||
+		s.rc.currProp.keys >= s.rc.propKeysIdxDistance {
+		newProp := *s.rc.currProp
+		s.rc.props = append(s.rc.props, &newProp)
+
+		s.rc.currProp.key = nil
+		s.rc.currProp.offset = s.offset
+		s.rc.currProp.keys = 0
+		s.rc.currProp.size = 0
+	}
+
+	return nil
+}
+
+var statSuffix = filepath.Join("_stat", "0")
+
+// GetAllFileNames returns a FilePathHandle that contains all data file paths
+// and a slice of stat file paths.
+func GetAllFileNames(
+	ctx context.Context,
+	store storage.ExternalStorage,
+	subDir string,
+) (FilePathHandle, []string, error) {
+	var dataFilePaths FilePathHandle
+	var stats []string
+
+	err := store.WalkDir(ctx,
+		&storage.WalkOption{SubDir: subDir},
+		func(path string, size int64) error {
+			if strings.HasSuffix(path, statSuffix) {
+				stats = append(stats, path)
+			} else {
+				dir, file := filepath.Split(path)
+				writerID, err := strconv.Atoi(filepath.Base(dir))
+				if err != nil {
+					return err
+				}
+				seq, err := strconv.Atoi(file)
+				if err != nil {
+					return err
+				}
+				dataFilePaths.set(writerID, seq, path)
+			}
+			return nil
+		})
+	if err != nil {
+		return dataFilePaths, nil, err
+	}
+	return dataFilePaths, stats, nil
+}
+
+// FilePathHandle handles data file paths under a prefix path.
+type FilePathHandle struct {
```

Comment:
Maybe we can abandon this structure and use `[]string` instead.

Reply:
I will think again when port the usage of FilePathHandle